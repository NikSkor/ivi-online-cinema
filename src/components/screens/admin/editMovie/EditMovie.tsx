import { FC, useEffect, useState } from "react";
import BreadCrumbs from '@/components/ui/breadCrumbs/BreadCrumbs';
import style from './EditMovie.module.scss';
import { IFilmographyItem } from "@/interfaces/person/IFilmographyItem";
import { filmographyData } from "@/components/ui/filmography/filmography.data";
import { useAppDispatch, useAppSelector } from "@/store/hooks";
import Link from "next/link";
import { addFilmValues } from '@/store/slices/adminSlice';


interface IFilmItem {
  name: string,
  foreignName: string,
  posterURL: string,
  year: number,
  rating: number,
}

const EditMovie: FC = () => {

  const id: number = useAppSelector(state => state.admin.filmId);

  const dispatch = useAppDispatch();


  let filmItem: IFilmItem = {
    name: '',
    foreignName: '',
    posterURL: '',
    year: 0,
    rating: 0,
  }

  let catalog: IFilmographyItem[] = [...filmographyData];
  
  if (id !== undefined) {
    catalog.forEach((item) => {
    if (item.movieId === +id) {
      Object.assign(filmItem, item);
      dispatch(addFilmValues(filmItem));
    }
  })
  }

  let [name, setName] = useState(filmItem.name);
  let [foreignName, setForeignName] = useState(filmItem.foreignName);
  let [urlImg, setUrlImg] = useState(filmItem.posterURL);
  let [year, setYear] = useState(filmItem.year);
  let [rating, setRating] = useState(filmItem.rating);

  let resetHandler = (e: any) => {
    e.preventDefault();

    setName(filmItem.name);
    setForeignName(filmItem.foreignName);
    setUrlImg(filmItem.posterURL);
    setYear(filmItem.year);
    setRating(filmItem.rating);

  }

  let submitHandler = (e: any) => {
    e.preventDefault();

    let filmValues: IFilmItem = {
    name: name,
    foreignName: foreignName,
    posterURL: urlImg,
    year: year,
    rating: rating,
  }
  dispatch(addFilmValues(filmValues));
  }

  let foreignNameHandler = (e: any) => {
    let reg = /[а-яА-ЯёЁ]/g;
    if (e.target.value.search(reg) !=  -1) {
        e.target.value  =  e.target.value.replace(reg, '');
    }
    setForeignName(e.target.value)
  }

  
  return(
      <div className="container">
        <section className={style.header}>
          <BreadCrumbs 
            pathList={[{pathLink: '/admin', pathName: 'Администратор'}]} 
            slug={'Фильм'} /> 
        </section>
        <section className={style.main}>
          <h2 className={style.title}>{name}</h2>
          <div className={style.form}>
            <label className={style.label} data-id='url'>
            Ссылка на постер:
            <input 
              disabled 
              className={style.inputs} 
              type='text'
              value={urlImg} 

              onChange={(e) => {setUrlImg(e.target.value)}}/>
          </label>
          <label className={style.label} data-id='url'>
            Название:
            <input placeholder="Введите название" className={style.inputs} type='text' value={name} onChange={(e) => {setName(e.target.value)}}/>
          </label>
          <label className={style.label} data-id='url'>
            Название на английском:
            <input placeholder="Введите название" className={style.inputs} type='text' value={foreignName} onChange={(e) => {foreignNameHandler(e)}}/>
          </label>
          <label className={style.label} data-id='url'>
            Год:
            <input disabled className={style.inputs} type='text' value={year} onChange={(e) => {setYear(+e.target.value)}}/>
          </label>
          <label className={style.label} data-id='url'>
            Рейтинг:
            <input disabled className={style.inputs} type='number' value={rating} onChange={(e) => {setRating(+e.target.value)}}/>
          </label>
          </div>
          <button className={style.actionBtn} onClick={(e) => {resetHandler(e)}}>Сбросить</button>
          <button className={style.actionBtn} onClick={(e) => {submitHandler(e)}}>Сохранить</button>
          <Link href={'/admin'} onClick={() => {
            }}>
          <button className={style.actionBtn}>
            Назад
          </button>
        </Link>
        </section>
      </div>
  )
}

export default EditMovie;