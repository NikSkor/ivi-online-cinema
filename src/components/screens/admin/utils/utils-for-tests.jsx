import React from "react";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import adminReducer from '../../../../store/slices/adminSlice';


export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = configureStore({
      reducer: { admin: adminReducer},
      preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}