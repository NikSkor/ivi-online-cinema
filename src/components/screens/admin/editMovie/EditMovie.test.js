import { cleanup, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import EditMovie from '@/components/screens/admin/EditMovie/EditMovie';
import { renderWithProviders } from '../utils/utils-for-tests';
import mockRouter from 'next-router-mock';


jest.mock('next/router', () => require('next-router-mock'));

describe('test EditGenre component', () => {
  afterEach(cleanup);

  test('there are the arrow btns', () => {
    renderWithProviders(<EditMovie />);
    const btns = screen.getAllByRole('button');
    btns.forEach((item) => {
      expect(item).toBeInTheDocument();
      expect(item).toMatchSnapshot();
    });
  });

  test('extBtn returns to admin page', () => {
    mockRouter.push('/admin');

    renderWithProviders(<EditMovie />);

    fireEvent.click(screen.getByText('Back'));

    expect(mockRouter).toMatchObject({
      asPath: '/admin',
      pathname: '/admin',
    });
  });
});