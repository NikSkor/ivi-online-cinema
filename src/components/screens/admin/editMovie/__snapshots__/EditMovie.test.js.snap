// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test EditGenre component there are the arrow btns 1`] = `
<button
  class="actionBtn"
>
  Reset
</button>
`;

exports[`test EditGenre component there are the arrow btns 2`] = `
<button
  class="actionBtn"
>
  Save
</button>
`;

exports[`test EditGenre component there are the arrow btns 3`] = `
<button
  class="actionBtn"
>
  Back
</button>
`;

exports[`test EditGenre component there are the arrow btns 4`] = `
<button
  class="btnClose"
>
  Закрыть
</button>
`;
